lms/djangoapps/courseware/masquerade.py

```diff
@@ -116,3 +179,83 @@ def get_masquerading_group_info(user, course_key):
     if not course_masquerade:
         return None, None
     return course_masquerade.group_id, course_masquerade.user_partition_id
+
+
+# Sentinel object to mark deleted objects in the session cache
+_deleted_sentinel = object()
+
+
+class MasqueradingKeyValueStore(KeyValueStore):
+    """
+    A `KeyValueStore` to avoid affecting the user state when masquerading.
+
+    This `KeyValueStore` wraps an underlying `KeyValueStore`.  Reads are forwarded to the underlying
+    store, but writes go to a Django session (or other dictionary-like object).
+    """
+    def __init__(self, kvs, session):
+        """
+        Arguments:
+          kvs: The KeyValueStore to wrap.
+          session: The Django session used to store temporary data in.
+        """
+        self.kvs = kvs
+        self.session = session
+        self.session_data = session.setdefault(MASQUERADE_DATA_KEY, {})
+
+    def _serialize_key(self, key):
+        """
+        Convert the key of Type KeyValueStore.Key to a string.
+
+        Keys are not JSON-serializable, so we can't use them as keys for the Django session.
+        The implementation is taken from cms/djangoapps/contentstore/views/session_kv_store.py.
+        """
```

Comment:
Is there enough code in common between this and `session_kv_store.py` to warrant moving it into either `openedx/` or `common/lib/xmodule/`?